### ISSUE TITLE
Azure Image Builder extension request - API 2022-07-01

### ISSUE THREAD

**Issue by zdelacerda-microsoft** [Azure/azure-cli#25853]
### Related command

Triggers (4 new commands, new sub group "trigger", similar to customizer, identity, validator, etc) 
*this is under a feature flag

1. `Az image builder trigger create  --trigger-name --resource-group --type (options: source) --image template` creates a new trigger to image template
2. `Az image builder trigger delete --trigger-name --resource-group --image template ` deletes specified trigger from template
3. `Az image builder trigger show --trigger-name --resource-group ` shows triggers available
4. `Az image builder trigger list --trigger-name --resource-group --image template` lists triggers in the template

____

Prefetch (3 new commands, also a new "optimize" sub category similar to customizer functionality)

1. `az image builder optimizer add --optimizer-name --type {vmBoot} --state {enabled:disabled}` add optimizer to an image template
2. `az image builder optimizer clear` removes all optimizers
3. `az image builder optimizer remove --optimizer-name` removes an optimizer
___

1. `az image builder output add --versioning (latest,source) --name` add versioning to specified output name with default latest or source options

2. `az image builder output versioning set --scheme (latest,source) --major {int} --name (output-name) ` add/change versioning options on output

3.`az image builder output versioning show -name (output-name)` show versioning options on output

4. `az image builder output versioning remove --name (output-name)` delete all versioning options on specified outputs

____
2 new Parameters:

1. `az image builder create --validator <file types>`

4. `az image builder output add --is-vhd --vhd-uri <string>`
*`--vhd-uri` is optional and requires `--is-vhd` to be used

___

### Resource Provider

Microsoft.VirtualMachineImages (Azure VM Image Builder)

### Description of Feature or Work Requested

We will now support:

1. Triggers - allows users to setup automation for image builds
2. New Validation Type "File" - allows files to be used for image validation
3. Prefetch/Optimizer - gives users the option to increase vm boot time by opting into this feature
4. Output versioning - allow customers to specify versioning scheme and associated options for their az image builder outputs

### Minimum API Version Required

2022-07-01

### Swagger PR link / SDK link

https://github.com/Azure/azure-rest-api-specs/pull/22901

### Request Example

Prefetch example:
https://msazure.visualstudio.com/One/_git/Compute-AzLinux-ImageBuilderRp?path=/swagger/api/2022-07-01/imagebuilder.server.json&version=GBdevelop&line=1503&lineEnd=1503&lineStartColumn=10&lineEndColumn=18&lineStyle=plain&_a=contents

### Target Date

ASAP as this is a few months behind

### Additional context

I (zdelacerda) will be oof 3/18-3/25. Please contact aibrpdevs@microsoft.com or kagarwal@microsoft.com if there are any follow up questions before my return. Thanks

**Comment by yonzhan** [comment]
compute feature

**Comment by zdelacerda-microsoft** [comment]
Hi AZ CLI reviewers, I will be oof 3/17-3/25. Please contact @feiyushi, @sumit-kalra or [aibrpdevs@microsoft.com](mailto:aibrpdevs@microsoft.com) for questions. Thank you

**Comment by zhoxing-ms** [comment]
Due to the big amount of content included in this feature request and our work at this sprint was already fully scheduled, we plan to release this feature in the next sprint **(on 04-25)**. May I ask do you think this release time is acceptable?

**Comment by feiyushi** [comment]
@zhoxing-ms 4/25 is fine. thanks.

**Comment by feiyushi** [comment]
Triggers commands group is in preview (behind AFEC "Microsoft.VirtualMachineImages/Triggers" please reach out to us in email for feature registration).

**Comment by zhoxing-ms** [comment]
> Triggers commands group is in preview (behind AFEC "Microsoft.VirtualMachineImages/Triggers" please reach out to us in email for feature registration).

May I ask is the status of this feature still private preview? Actually, we usually only build private package through edge build for private preview, and only release GA and public preview features in the official CLI version

**Comment by zhoxing-ms** [comment]
@zdelacerda-microsoft @feiyushi In addition, since the previous `az image builder` related command group were released in core CLI, do you want these command groups to also be released in core CLI or do you want us to create a new extension to release them?

**Comment by feiyushi** [comment]
@zhoxing-ms please exclude trigger command group from this release. We'll include it in the next one when we are in public preview/GA. Thanks!
@zdelacerda-microsoft

**Comment by zhoxing-ms** [comment]
@feiyushi OK, thanks for your clarification. Could you please add a new feature request for trigger command group to help us plan and follow up on it?

**Comment by yanzhudd** [comment]
> @zhoxing-ms please exclude trigger command group from this release. We'll include it in the next one when we are in public preview/GA. Thanks! @zdelacerda-microsoft

Hi @feiyushi 
Please let me check if the following triiger group could be ignored in this feature request?

> 
> Triggers (4 new commands, new sub group "trigger", similar to customizer, identity, validator, etc)
> *this is under a feature flag
> 
> Az image builder trigger create  --trigger-name --resource-group --type (options: source) --image template creates a new trigger to image template
> Az image builder trigger delete --trigger-name --resource-group --image template  deletes specified trigger from template
> Az image builder trigger show --trigger-name --resource-group  shows triggers available
> Az image builder trigger list --trigger-name --resource-group --image template lists triggers in the template

**Comment by feiyushi** [comment]
@yanzhudd please exclude trigger cmd group from this release. thanks.

**Comment by zhoxing-ms** [comment]
@feiyushi Could you please remove the `trigger cmd group` part from the current issue and add them to a separate new issue? This is convenient for us to follow up and plan them according to their different release times

**Comment by yanzhudd** [comment]
> Prefetch (3 new commands, also a new "optimize" sub category similar to customizer functionality)
> 
> az image builder optimizer add --optimizer-name --type {vmBoot} --state {enabled:disabled} add optimizer to an image template
> az image builder optimizer clear removes all optimizers
> az image builder optimizer remove --optimizer-name removes an optimizer

Hi @feiyushi @zdelacerda-microsoft 
since this prefetch feature is realized by updating a parameter `ImageTemplateProperties.optimize` of an image template, does the format of these new commands need to be changed to:

az image builder optimizer add **_--image-template-name_** --type {vmBoot} --state {enabled:disabled}: add optimizer to an image template
az image builder optimizer remove **_--image-template-name_**: removes an optimizer

The `clear` command here is removed as the `ImageTemplateProperties.optimize` parameter is not a list and `remove` command seems to be sufficient.

Could you please help to confirm this?

**Comment by yanzhudd** [comment]
Hi @feiyushi @zdelacerda-microsoft 
When we develop the feature of `az image builder optimizer` group, an error is reported. Could you please help to check the reason?
![image](https://user-images.githubusercontent.com/105691024/229727460-1e537cc3-ba9c-4731-aa59-3ccb3042d698.png)

**Comment by yanzhudd** [comment]
Hi @zdelacerda-microsoft 
could you please provide more details of this feature, for example the backend parameters of the new commands and new arguments?

**Comment by zdelacerda-microsoft** [comment]
@yanzhudd yes i can. specifically which commands are you asking about?

**Comment by yanzhudd** [comment]
> @yanzhudd yes i can. specifically which commands are you asking about?

for example, the details including:

- > az image builder output add --versioning (latest,source) --name add versioning to specified output name with default latest or source options

    what is **backend parameter** of `--versioning (latest,source)` in this command?

- what is the **request path** and **parameters** like `--scheme (latest,source)` `--major {int}` for `az image builder output versioning` command?

- what is the **backend parameter** for the new added parameter `--validator <file types>` in `az image builder create` command?

- what is the **backend parameters** for the new added parameters `--is-vhd` `--vhd-uri <string>` in `az image builder output add` command?

thanks for clarification.

**Comment by uiri** [comment]
@yanzhudd 

Regarding `versioning`, the JSON representation of the builder output for the REST API is like this:

```
      {
        "type": "SharedImage",
        "galleryImageId": "/subscriptions/SUBSCRIPTION/resourceGroups/TEST_RG/providers/Microsoft.Compute/galleries/tes\
t_image_gallery/images/ubuntu",
        "runOutputName": "sig-ubuntu-1604",
        "replicationRegions": [
          "westus2"
        ],
        "versioning": {
          "scheme": "Latest"
        }
      },
```

`scheme` is treated as an enum. Current valid values are `"Latest"` or `"Source"`.
Currently the only other possible property in `versioning` is `major` which is only valid for Latest and is an integer.

**Comment by zdelacerda-microsoft** [comment]
Does this answer your question @yanzhudd. Are we unblocked?

**Comment by yanzhudd** [comment]
Hi @zdelacerda-microsoft

it seems that the [SDK](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/compute/azure-mgmt-imagebuilder/azure/mgmt/imagebuilder/models/_models_py3.py) is not ready for this feature since the parameters `optimize`, `versioning` and `uri` of `ImageTemplateVhdDistributor` do not exist.

Please refer to the [SDK release doc](https://github.com/Azure/azure-rest-api-specs/wiki/Abstract-about-how-to-ask-for-public-SDK-release) to release a new version of SDK firstly, so that we CLI could develop this feature.

Mentioned in the SDK release doc:

> In short, if you make release issue before 10th every month, it will be released before 24th; if later than 10th, the release will be delayed to next month.

Our CLI is developed depending on SDK. The earliest time of SDK release is next month, so this feature has to be postponed to next sprint (05-23). If anything urgent, we could provide private package by edge build before our official release.

**Comment by yanzhudd** [comment]
> az image builder create --validator <file types>

In addition, may I ask what is the backend property of `--validator`?

**Comment by zdelacerda-microsoft** [comment]
@yanzhudd which SDK are you using? all of our SDKs (go, java, js, and python) are completed and are being tested right now.

**Comment by zdelacerda-microsoft** [comment]
This is the backend for --validator

ImageTemplateInVMValidator": {
      "type": "object",
      "description": "Describes a unit of in-VM validation of image",
      "discriminator": "type",
      "properties": {
        "type": {
          "type": "string",
          "description": "The type of validation you want to use on the Image. For example, \"Shell\" can be shell validation"
        },
        "name": {
          "type": "string",
          "description": "Friendly Name to provide context on what this validation step does"
        }
      },
      "required": [
        "type"

**Comment by yanzhudd** [comment]
> @yanzhudd which SDK are you using? all of our SDKs (go, java, js, and python) are completed and are being tested right now.

we are using python SDK. It seems the new added parameters do not exist in the main branch (https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/compute/azure-mgmt-imagebuilder/azure/mgmt/imagebuilder/models/_models_py3.py)

> it seems that the [SDK](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/compute/azure-mgmt-imagebuilder/azure/mgmt/imagebuilder/models/_models_py3.py) is not ready for this feature since the parameters optimize, versioning and uri of ImageTemplateVhdDistributor do not exist.

**Comment by zdelacerda-microsoft** [comment]
@yanzhudd our python SDK has now been updated. https://pypi.org/project/azure-mgmt-imagebuilder/1.2.0/

when can we expect our az cli to be updated?

**Comment by yanzhudd** [comment]
@zdelacerda-microsoft if the process and test go well, this feature will be updated in the next release of CLI (5/23).

**Comment by yanzhudd** [comment]
> Prefetch (3 new commands, also a new "optimize" sub category similar to customizer functionality)
> 
> az image builder optimizer add --optimizer-name --type {vmBoot} --state {enabled:disabled} add optimizer to an image template
> az image builder optimizer clear removes all optimizers
> az image builder optimizer remove --optimizer-name removes an optimizer

Hi @zdelacerda-microsoft 
In `az image builder optimizer` group, `--optimizer-name` seems to have a same meaning as `--type` (we dont know if we understand it correctly, if not, please let us know). May I ask whether it is possible to remove the `--optimizer-name` parameter? 

And to keep consistent with `add` command, we thought it is better to add `--type {vmBoot}` parameter to `remove` command to specify which type of optimizer would be removed. Could you please help to confirm if the commands of `az image builder optimizer` group listed here meet your expectation?

`az image builder optimizer add --resource-group --image-template-name --type {vmBoot} --state {enabled:disabled}` adds optimizer to an image template
`az image builder optimizer clear --resource-group --image-template-name` removes all optimizers
`az image builder optimizer remove --resource-group --image-template-name --type {vmBoot}` removes optimizer from an image template

**Comment by yanzhudd** [comment]
> @yanzhudd please exclude trigger cmd group from this release. thanks.

Hi @feiyushi may I ask whether it is ready to release trigger group in this sprint (5/23)?

**Comment by yanzhudd** [comment]
> This is the backend for --validator
> 
> ImageTemplateInVMValidator": { "type": "object", "description": "Describes a unit of in-VM validation of image", "discriminator": "type", "properties": { "type": { "type": "string", "description": "The type of validation you want to use on the Image. For example, "Shell" can be shell validation" }, "name": { "type": "string", "description": "Friendly Name to provide context on what this validation step does" } }, "required": [ "type"

@zdelacerda-microsoft does `--validator` parameter correspond to `type` with string type of `ImageTemplateInVMValidator class`?

**Comment by feiyushi** [comment]
> 

@yanzhudd we plan to release a public preview for the trigger feature so can you complete the trigger group in this release 5/23?
we can open up the registration for the feature "Microsoft.VirtualMachineImages/Triggers" soon if you need it for development.

**Comment by yanzhudd** [comment]
> > 
> 
> @yanzhudd we plan to release a public preview for the trigger feature so can you complete the trigger group in this release 5/23? we can open up the registration for the feature "Microsoft.VirtualMachineImages/Triggers" soon if you need it for development.

@feiyushi okay, I will complete the triiger group in this sprint if test works well.
Do you need our subId for registration?

**Comment by yanzhudd** [comment]
> @sumit-kalra can help answer any questions about validator

Since we will have a code freeze next Friday (5-19) for the public release. Please let us know the info of `--validator` parameter in `az image builder create` command before 5-15, so that we could have time to develop and test. Otherwise, it has to be postponed to **next sprint (7-4)**.

**Comment by zhoxing-ms** [comment]
@zdelacerda-microsoft Please pay attention to this comment https://github.com/Azure/azure-cli/pull/26261#discussion_r1196256914

**Comment by zdelacerda-microsoft** [comment]
@yanzhudd does next spring (7-4) mean eta is now July 4th?

**Comment by yanzhudd** [comment]
> @yanzhudd does next spring (7-4) mean eta is now July 4th?

yes. We will release the part of code in 7-4.

**Comment by sumit-kalra** [comment]
> > Since we will have a code freeze next Friday (5-19) for the public release. Please let us know the info of `--validator` parameter in `az image builder create` command before 5-15, so that we could have time to develop and test. Otherwise, the change of this command has to be postponed to **next sprint (7-4)**.
> 
> @sumit-kalra please reply to this message

Yes, a single instance of validator corresponds to `ImageTemplateInVMValidator`.

However, `ImageTemplateInVMValidator` itself is embedded in `validate` - which has couple other properties:
1. `continueDistributeOnFailure`
2. `sourceValidationOnly`
Are these two properties handled elsewhere?

**Comment by yanzhudd** [comment]
> 

yes, `ImageTemplatePropertiesValidate` and `ImageTemplatePropertiesValidate` are handled in `az image builder validator add` command.

**Comment by yanzhudd** [comment]
> > > Since we will have a code freeze next Friday (5-19) for the public release. Please let us know the info of `--validator` parameter in `az image builder create` command before 5-15, so that we could have time to develop and test. Otherwise, the change of this command has to be postponed to **next sprint (7-4)**.
> > 
> > 
> > @sumit-kalra please reply to this message
> 
> Yes, a single instance of validator corresponds to `ImageTemplateInVMValidator`.
> 
> However, `ImageTemplateInVMValidator` itself is embedded in `validate` - which has couple other properties:
> 
> 1. `continueDistributeOnFailure`
> 2. `sourceValidationOnly`
>    Are these two properties handled elsewhere?

Hi @sumit-kalra, 

since there is a command group `az image builder validator` to manage image builder template validate, it is suggested to add the parameter `--type` to the `az image builder validator add` command intead of adding `--validator` to `az image builder create` command, i.e.,

`az image builder validator add -n myTemplate -g myGroup --continue-distribute-on-failure true --source-validation-only true --type shell --defer`

Could you please help to confirm if it meets your expectation?

**Comment by yanzhudd** [comment]
> @yanzhudd @zhoxing-ms - it looks like our prefetch commands are missing from the az cli docs. were these commands implemented?
> 
> Prefetch (3 new commands, also a new "optimize" sub category similar to customizer functionality)
> 
> az image builder optimizer add --optimizer-name --type {vmBoot} --state {enabled:disabled} add optimizer to an image template az image builder optimizer clear removes all optimizers az image builder optimizer remove --optimizer-name removes an optimizer
> 
> https://learn.microsoft.com/en-us/cli/azure/image/builder?view=azure-cli-latest

@zdelacerda-microsoft 
this group of command needed to be confirmed the format last sprint (https://github.com/Azure/azure-cli/issues/25853#issuecomment-1551136199 https://github.com/Azure/azure-cli/pull/26261#discussion_r1196256914) and will be released in this sprint (7-4).

**Comment by zdelacerda-microsoft** [comment]
hi @yanzhudd @zhoxing-ms are we still on track for 7/4 release for the rest of our commands?

**Comment by yanzhudd** [comment]
Hi @zdelacerda-microsoft 
yes, the rest of the commands will be released in 7-4 with a new version of CLI.
